applications/GeoMechanicsApplication/tests/cpp_tests/test_time_incrementor.cpp

```diff
@@ -0,0 +1,439 @@
+    const auto end_time          = start_time + 8.0;
+    const auto start_increment   = 0.5;
+    const auto time_incrementor  = AdaptiveTimeIncrementor{start_time, end_time, start_increment};
+    auto       previous_state    = TimeStepEndState{};
+    previous_state.time          = end_time;
+
+    KRATOS_EXPECT_FALSE(time_incrementor.WantNextStep(previous_state))
+}
+
+KRATOS_TEST_CASE_IN_SUITE(RetryWhenNotAttemptedYet, KratosGeoMechanicsFastSuite)
+{
+    const auto start_time        = 0.0;
+    const auto end_time          = start_time + 8.0;
+    const auto start_increment   = 0.5;
+    const auto max_num_of_cycles = std::size_t{4};
+    const auto time_incrementor  = AdaptiveTimeIncrementor{start_time, end_time, start_increment, max_num_of_cycles};
+    const auto cycle_number      = std::size_t{0};
+    const auto previous_state    = TimeStepEndState{};
+
+    KRATOS_EXPECT_TRUE(time_incrementor.WantRetryStep(cycle_number, previous_state))
+}
+
+KRATOS_TEST_CASE_IN_SUITE(RetryWhenPreviousAttemptDidNotConvergeButAtLeastOneMoreLeft, KratosGeoMechanicsFastSuite)
+{
+    const auto start_time        = 0.0;
+    const auto end_time          = start_time + 8.0;
+    const auto start_increment   = 0.5;
+    const auto max_num_of_cycles = std::size_t{4};
+    const auto time_incrementor  = AdaptiveTimeIncrementor{start_time, end_time, start_increment, max_num_of_cycles};
+    const auto cycle_number      = max_num_of_cycles - 1;
+    auto previous_state          = TimeStepEndState{};
+    previous_state.convergence_state = TimeStepEndState::ConvergenceState::non_converged;
+
+    KRATOS_EXPECT_TRUE(time_incrementor.WantRetryStep(cycle_number, previous_state))
+}
+
+KRATOS_TEST_CASE_IN_SUITE(DontRetryWhenPreviousAttemptDidNotConvergeAndNoAttemptsLeft, KratosGeoMechanicsFastSuite)
+{
+    const auto start_time        = 0.0;
+    const auto end_time          = start_time + 8.0;
+    const auto start_increment   = 0.5;
+    const auto max_num_of_cycles = std::size_t{4};
+    const auto time_incrementor  = AdaptiveTimeIncrementor{start_time, end_time, start_increment, max_num_of_cycles};
+    auto previous_state          = TimeStepEndState{};
+    previous_state.convergence_state = TimeStepEndState::ConvergenceState::non_converged;
+
+    KRATOS_EXPECT_FALSE(time_incrementor.WantRetryStep(max_num_of_cycles, previous_state))
+}
+
+KRATOS_TEST_CASE_IN_SUITE(DontRetryWhenPreviousAttemptConverged, KratosGeoMechanicsFastSuite)
+{
+    const auto start_time        = 0.0;
+    const auto end_time          = start_time + 8.0;
+    const auto start_increment   = 0.5;
+    const auto max_num_of_cycles = std::size_t{4};
+    const auto time_incrementor  = AdaptiveTimeIncrementor{start_time, end_time, start_increment, max_num_of_cycles};
+    const auto cycle_number      = std::size_t{1};
+    auto previous_state          = TimeStepEndState{};
+    previous_state.convergence_state = TimeStepEndState::ConvergenceState::converged;
```

Comment:
Test is correct, but the practical fill in is unexpected. cycle_Number == 1, a previous cycle does is sort of a starting state which I expect to have convergence_state NonConverged. Here it is set to Converged. Maybe set cycle_number > 1 to avoid this confusion.

Reply:
Cycle number 0 is the first attempt, where it may or may not converge. The test reflects a case where convergence was achieved after this first attempt. As discussed, the confusion is due an off-by-one error.